src/Test/StateMachine.hs

```diff
@@ -70,47 +90,51 @@ forAllProgram generator shrinker precondition transition model =
     (evalStateT (generateProgram generator precondition transition 0) model)
     (shrinkProgram shrinker precondition transition model)
 
--- | Run a sequential program and check if your model agrees with your
---   semantics.
-runAndCheckProgram
+monadicSequential
   :: Monad m
-  => HFunctor act
-  => Precondition model act
-  -> Transition model act
-  -> Postcondition model act
-  -> InitialModel model
-  -> Semantics act m
-  -> (m Property -> Property)  -- ^ Runner
-  -> Program act
+  => Show (Untyped act)
+  => HFoldable act
+  => StateMachine model act m
+  -> (Program act -> PropertyM m a)
   -> Property
-runAndCheckProgram precond trans postcond m sem runner =
-  runAndCheckProgram' precond trans postcond m sem (return ()) (const runner) (const (return ()))
+monadicSequential StateMachine {..} predicate
+  = forAllProgram generator' shrinker' precondition' transition' model'
+  $ monadic (ioProperty . runner')
+  . predicate
 
--- | Same as above, except with the possibility to setup some resource
---   for the runner to use. The resource could be a database connection
---   for example.
-runAndCheckProgram'
-  :: Monad m
+runProgram
+  :: forall m act model
+  .  Monad m
+  => Show (Untyped act)
   => HFunctor act
-  => Precondition model act
-  -> Transition model act
-  -> Postcondition model act
-  -> InitialModel model
-  -> Semantics act m
-  -> IO setup                           -- ^ Setup a resource.
-  -> (setup -> m Property -> Property)
-  -> (setup -> IO ())                   -- ^ Tear down the resource.
+  => StateMachine  model act m
   -> Program act
-  -> Property
-runAndCheckProgram' precond trans postcond m sem setup runner cleanup acts =
-  monadic (ioProperty . runnerWithSetup)
-    (checkProgram precond trans postcond m m sem acts)
+  -> PropertyM m (History act, model Concrete, Property)
+runProgram sm = run . executeProgram sm
+
+prettyProgram
+  :: MonadIO m
+  => Program act -> History act -> model Concrete -> Property -> PropertyM m ()
+prettyProgram _ hist _ prop = putStrLn (ppHistory hist) `whenFailM` prop
+
+actionNames :: forall act. Show (Untyped act) => Program act -> [(String, Int)]
```

Comment:
Similar here we can use `Data.Data` to get the names of the constructors instead of using `Show`

```haskell
  name = show . toConstr
```

Reply:
Of course that means that all examples will have to derive `Data`, and hmm that seems complicated for GADTs..

Reply:
Yeah... I just tried and deriving `Data` fails. Same with `Generic` which would have been another option, so I guess TH it is?